vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs

```diff
@@ -69,30 +70,91 @@ type internal FSharpGoToDefinitionService
+                                            checkResults.GetAllUsesOfAllSymbolsInFile(cancellationToken)
+                                            |> Seq.tryFind (fun x ->
+                                                match x.Symbol, targetSymbolUse.Symbol with
+                                                | (:? FSharpEntity as symbol1), (:? FSharpEntity as symbol2) when x.IsFromDefinition ->
+                                                    symbol1.DisplayName = symbol2.DisplayName
+                                                | (:? FSharpMemberOrFunctionOrValue as symbol1), (:? FSharpMemberOrFunctionOrValue as symbol2) ->
+                                                    symbol1.DisplayName = symbol2.DisplayName &&
+                                                    symbol1.GenericParameters.Count = symbol2.GenericParameters.Count &&
+                                                    symbol1.CurriedParameterGroups.Count = symbol2.CurriedParameterGroups.Count &&
+                                                    (
+                                                        (symbol1.CurriedParameterGroups, symbol2.CurriedParameterGroups)
+                                                        ||> Seq.forall2 (fun pg1 pg2 ->
+                                                            pg1.Count = pg2.Count &&
+                                                            (
+                                                                (pg1, pg2)
+                                                                ||> Seq.forall2 (fun p1 p2 ->
+                                                                    areTypesEqual p1.Type p2.Type
+                                                                )
+                                                            )
+                                                        )
+                                                    ) &&
+                                                    areTypesEqual symbol1.ReturnParameter.Type symbol2.ReturnParameter.Type
+                                                | _ ->
+                                                    false
+                                            )
+                                            |> Option.map (fun x -> x.Range)
+
+                                        let span =
+                                            match RoslynHelpers.TryFSharpRangeToTextSpan(tmpShownDoc.GetTextAsync(cancellationToken).Result, r) with
+                                            | Some span -> span
+                                            | _ -> TextSpan()
+
+                                        return span                             
+                                    }
+
+                                let span =
+                                    match Async.RunSynchronously(goToAsync, cancellationToken = cancellationToken) with
```

Comment:
This has my ears raised a bit. Is there an awaitable thing we could potentially plug this into? Naiive question and I know that this whole thing runs async anyways, but I fear that _something_ in VS will cause us to run in the UI thread and we'd introduce a UI freeze. I don't have anything concrete to test this with though, and so my concern there might be meaningless.

Reply:
TryGoToDefinition is not awaitable, even in Roslyn from last I checked. It must block, but while it's blocked it will show the animating status bar which is cancellable.

The calls made in the `asyncMaybe` block don't call into anything that requires the UI thread, it just calls FCS async functions and gets the source text from the document.